framework/validators/IpValidator.php

```diff
@@ -0,0 +1,555 @@
+
+    /**
+     * @var boolean whether the address can be a CIDR subnet
+     *    true - the subnet is required
+     *   false - the address can not have the subnet
+     *    null - ths subnet is optional
+     */
+    public $subnet = false;
+
+    /**
+     * @var boolean whether to add the prefix with the smallest length (32 for IPv4 and 128 for IPv6)
+     * to an address without it.
+     * Works only when attribute [[subnet]] is not false.
+     * Example: `10.0.1.5` will normalized to `10.0.1.5/32`, `2008:db0::1` will be normalized to `2008:db0::1/128`
+     * @see subnet
+     */
+    public $normalize = false;
+
+    /**
+     * @var string|boolean
+     * @see negationChar
+     */
+    public $_negationChar = false;
+
+    /**
+     * @var boolean whether to expand an IPv6 address to the full notation format
+     */
+    public $expandV6 = false;
+
+    /**
+     * The property is used by [[isAllowed]] method to determine the order of checks
+     * of the IP address according to [[deny]] and [[allow]] lists.
+     *
+     * When [[order]] is [[ORDER_ALLOW_DENY]] - first all the [[allow]] will be checked, and at least one must
+     * match or the validation will fail. Then all the [[deny]] are checked, if one of them
+     * matched the validation will fail.
+     * At last, if address is not on the [[allow]] nor on the [[deny]] the validation will also fail.
+     *
+     * When [[order]] is [[ORDER_DENY_ALLOW]] - method [[isAllowed]] checks all the [[deny]] and the [[allow]].
+     * If the value is on the [[deny]] the validation will fail, unless it is also present on the [[allow]].
+     * If it is not found on any of the lists the validation will be passed.
+     *
+     * Tip: it is useful to use [[ORDER_ALLOW_DENY]], when it is necessary to deny a less specific subnet and
+     * to allow a more specific one. The example below will result in passing `192.168.1.1`,
+     * but `192.168.2.1` will be denied:
+     *
+     * ```
+     * [
+     *      'deny' => ['192.168.0.0/16'],
+     *      'allow' => ['192.168.1.0/24']
+     * ]
+     * ```
+     *
+     * @var int the order of ranges rules. Used by [[isAllowed]]
+     * @see isAllowed()
+     */
+    public $order = self::ORDER_DENY_ALLOW;
+
+    /**
+     * @var string|array IPv4 or IPv6 ranges that are allowed to use. For example:
+     *
+     * ```
+     * ['10.0.0.0/8', '192.168.1.1', '2001:ab::/64', '2001:ac::2:1']
+     * ```
+     *
+     * @see isAllowed()
+     */
+    public $allow;
+
+    /**
+     * @var string|array IPv4 or IPv6 ranges that are prohibited to use. For example:
+     *
+     * ```
+     * ['10.0.0.0/24', '192.168.2.1', '2001:ab::/32', '2001:ac::1:2']
+     * ```
+     *
+     * @see isAllowed()
+     */
+    public $deny;
+
+    /**
+     * @var string Regexp-pattern to validate IPv4 address
+     */
+    public $ipv4Pattern = '/^(?:(?:2(?:[0-4][0-9]|5[0-5])|[0-1]?[0-9]?[0-9])\.){3}(?:(?:2([0-4][0-9]|5[0-5])|[0-1]?[0-9]?[0-9]))$/';
```

Comment:
`[0-9]` could be replaced with `\d` in order to shorten it a bit. Not critical though...